File: Motocitizen/src/motocitizen/startup/Preferences.java
package motocitizen.startup;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.SharedPreferences;
import android.media.RingtoneManager;
import android.net.Uri;
import android.preference.PreferenceManager;

import com.google.android.gms.maps.model.LatLng;

import org.json.JSONArray;
import org.json.JSONException;

import java.util.Arrays;

import motocitizen.content.Type;
import motocitizen.main.R;
import motocitizen.utils.Const;

@SuppressLint("CommitPrefEdits")
public class Preferences {

    public final static  String showAcc             = "mc.show.acc";
    public final static  String showBreak           = "mc.show.break";
    public final static  String showSteal           = "mc.show.steal";
    public final static  String showOther           = "mc.show.other";
    public final static  String distanceShow        = "mc.distance.show";
    public final static  String distanceAlarm       = "mc.distance.alarm";
    public final static  String mapProvider         = "mc.map.provider";
    public final static  String currentVersion      = "version";
    public final static  String doNotDisturb        = "do.not.disturb";
    public final static  String hoursAgo            = "hours.ago";
    public final static  String maxNotifications    = "notifications.max";
    public final static  String useVibration        = "use.vibration";
    public final static  String userId              = "userId";
    public final static  String userName            = "userName";
    public final static  String userRole            = "userRole";
    private final static String onWay               = "mc.onway";
    private final static String soundTitle          = "mc.notification.sound.title";
    private final static String soundURI            = "mc.notification.sound";
    private final static String login               = "mc.login";
    private final static String password            = "mc.password";
    private final static String anonim              = "mc.anonim";
    private final static String GCMRegistrationCode = "mc.gcm.id";
    private final static String appVersion          = "mc.app.version";
    private final static String savedLng            = "savedlng";
    private final static String savedLat            = "savedlat";
    private final static String notificationList    = "notificationList";
    private final static String messageReadList     = "messageReadList";

    private final static String[] mapProviders = {"google", "osm", "yandex"};

    private static SharedPreferences preferences;
    private static Context           context;

    public Preferences(Context context) {
        preferences = PreferenceManager.getDefaultSharedPreferences(context);
        Preferences.context = context;
    }

    public static void putBoolean(String name, boolean value) {
        preferences.edit().putBoolean(name, value).commit();
    }

    public static int getOnWay() {
        return preferences.getInt(onWay, 0);
    }

    public static void setOnWay(int id) {
        preferences.edit().putInt(onWay, id).commit();
    }

    public static LatLng getSavedLatLng() {
        double lat = (double) preferences.getFloat(savedLat, 55.752295f);
        double lng = (double) preferences.getFloat(savedLng, 37.622735f);
        return new LatLng(lat, lng);
    }

    public static boolean getDoNotDisturb() {
        return preferences.getBoolean(doNotDisturb, false);
    }

    public static void setDoNotDisturb(boolean value) {
        preferences.edit().putBoolean(doNotDisturb, value).commit();
    }

    public static String getCurrentVersion() {
        return preferences.getString(currentVersion, context.getString(R.string.unknown_code_version));
    }

    public static void setCurrentVersion(String version) {
        preferences.edit().putString(currentVersion, version).commit();
    }

    public static void saveLatLng(LatLng latlng) {
        preferences.edit().putFloat(savedLat, (float) latlng.latitude).putFloat(savedLng, (float) latlng.longitude).commit();
    }

    public static int getVisibleDistance() {
        int distance;
        try {
            distance = preferences.getInt(distanceShow, 200);
        } catch (Exception e) {
            String distanceString = preferences.getString(distanceShow, "200");
            distance = Integer.parseInt(distanceString);
        }
        return distance;
    }

    public static void setVisibleDistance(int distance) {
        if (distance > Const.EQUATOR) {
            distance = Const.EQUATOR;
        }
        preferences.edit().putString(distanceShow, String.valueOf(distance)).commit();
    }

    public static int getAlarmDistance() {
        int distance;
        try {
            distance = preferences.getInt(distanceAlarm, 20);
        } catch (Exception e) {
            String distanceString = preferences.getString(distanceAlarm, "20");
            distance = Integer.parseInt(distanceString);
        }
        return distance;
    }

    public static void setAlarmDistance(int distance) {
        if (distance > Const.EQUATOR) {
            distance = Const.EQUATOR;
        }
        preferences.edit().putString(distanceAlarm, String.valueOf(distance)).commit();
    }

    public static String getAlarmSoundTitle() {
        return preferences.getString(soundTitle, "default system");
    }

    public static Uri getAlarmSoundUri() {
        Uri    uri;
        String uriString = preferences.getString(soundURI, "default");
        if (uriString.equals("default")) {
            uri = RingtoneManager.getActualDefaultRingtoneUri(context, RingtoneManager.TYPE_NOTIFICATION);
        } else uri = Uri.parse(uriString);
        return uri;
    }

    public static String getMapProvider() {
        return preferences.getString(mapProvider, "google");
    }

    public static void setMapProvider(String provider) {
        if (Arrays.asList(mapProviders).contains(provider)) {
            preferences.edit().putString(mapProvider, provider);
        }
    }

    public static String getLogin() {
        return preferences.getString(login, "");
    }

    public static void setLogin(String value) {
        preferences.edit().putString(login, value).commit();
    }

    public static String getPassword() {
        return preferences.getString(password, "");
    }

    public static void setPassword(String value) {
        preferences.edit().putString(password, value).commit();
    }

    public static boolean isAnonim() {
        return preferences.getBoolean(anonim, false);
    }

    public static void setAnonim(boolean value) {
        preferences.edit().putBoolean(anonim, value).commit();
    }

    public static String getGCMRegistrationCode() {
        return preferences.getString(GCMRegistrationCode, "");
    }

    public static void setGCMRegistrationCode(String code) {
        preferences.edit().putString(GCMRegistrationCode, code).commit();
    }

    public static int getAppVersion() {
        return preferences.getInt(appVersion, 0);
    }

    public static void setAppVersion(int code) {
        preferences.edit().putInt(appVersion, code).commit();
    }

    public static void setShowAcc(boolean show) {
        preferences.edit().putBoolean(showAcc, show).commit();
    }

    public static void setShowBreak(boolean show) {
        preferences.edit().putBoolean(showBreak, show).commit();
    }

    public static void setShowSteal(boolean show) {
        preferences.edit().putBoolean(showSteal, show).commit();
    }

    public static void setShowOther(boolean show) {
        preferences.edit().putBoolean(showOther, show).commit();
    }

    public static void setSoundAlarm(String title, Uri uri) {
        preferences.edit().putString(soundTitle, title).putString(soundURI, uri.toString()).commit();
    }

    public static void setDefaultSoundAlarm() {
        preferences.edit().putString(soundTitle, "default system").putString(soundURI, "default").commit();
    }

    public static void resetAuth() {
        preferences.edit().remove(login).remove(password).commit();
    }

    public static boolean isHidden(Type type) {
        switch (type) {
            case BREAK:
                return !toShowBreak();
            case MOTO_AUTO:
            case MOTO_MOTO:
            case MOTO_MAN:
            case SOLO:
                return !toShowAcc();
            case STEAL:
                return !toShowSteal();
            case OTHER:
            default:
                return !toShowOther();
        }
    }

    public static boolean toShowBreak() {
        return preferences.getBoolean(showBreak, true);
    }

    public static boolean toShowAcc() {
        return preferences.getBoolean(showAcc, true);
    }

    public static boolean toShowSteal() {
        return preferences.getBoolean(showSteal, true);
    }

    public static boolean toShowOther() {
        return preferences.getBoolean(showOther, true);
    }

    public static void resetPreferences() {
        preferences.edit().clear();
    }

    public static int getMaxNotifications() {
        return Integer.parseInt(preferences.getString(maxNotifications, "3"));
    }

    public static void setMaxNotifications(int code) {
        preferences.edit().putInt(maxNotifications, code).commit();
    }

    public static JSONArray getNotificationList() {
        JSONArray json;
        try {
            json = new JSONArray(preferences.getString(notificationList, "[]"));
        } catch (JSONException e) {
            e.printStackTrace();
            json = new JSONArray();
        }
        return json;
    }

    public static void setNotificationList(JSONArray json) {
        preferences.edit().putString(notificationList, json.toString()).commit();
    }

    public static JSONArray getMessageReadList() {
        JSONArray json;
        try {
            json = new JSONArray(preferences.getString(messageReadList, "[]"));
        } catch (JSONException e) {
            e.printStackTrace();
            json = new JSONArray();
        }
        return json;
    }

    public static void setMessageReadList(JSONArray json) {
        preferences.edit().putString(messageReadList, json.toString()).commit();
    }

    public static int getHoursAgo() {
        return Integer.parseInt(preferences.getString(hoursAgo, "24"));
    }

    public static void setHoursAgo(int hours) {
        preferences.edit().putString(hoursAgo, String.valueOf(hours)).commit();
    }

    public static boolean getVibration() {
        return preferences.getBoolean(useVibration, true);
    }

    public static void setVibration(boolean vibration) {
        preferences.edit().putString(useVibration, String.valueOf(vibration)).commit();
    }

    public static String getUserName() {
        return preferences.getString(userName, "");
    }

    public static void setUserName(String name) {
        preferences.edit().putString(userName, name).commit();
    }

    public static int getUserId() {
        return preferences.getInt(userId, 0);
    }

    public static void setUserId(int id) {
        preferences.edit().putInt(userId, id).commit();
    }

    public static String getUserRole() {
        return preferences.getString(userRole, "readonly");
    }

    public static void setUserRole(String role) {
        preferences.edit().putString(userRole, role).commit();
    }
}

File: Motocitizen/src/motocitizen/datasources/preferences/Preferences.kt
package motocitizen.datasources.preferences

import android.content.Context
import android.content.SharedPreferences
import android.media.RingtoneManager
import android.net.Uri
import android.preference.PreferenceManager
import com.google.android.gms.maps.model.LatLng
import motocitizen.MyApp
import motocitizen.datasources.preferences.Preferences.Stored.*
import motocitizen.dictionary.Type
import kotlin.properties.ReadWriteProperty
import kotlin.reflect.KProperty

object Preferences {

    private val preferences: SharedPreferences = PreferenceManager.getDefaultSharedPreferences(MyApp.context)

    var appVersion by PreferenceDelegate<Int>(APP_VERSION)
    var doNotDisturb by PreferenceDelegate<Boolean>(DO_NOT_DISTURB)
    var onWay by PreferenceDelegate<Int>(ON_WAY)
    var soundTitle by PreferenceDelegate<String>(SOUND_TITLE)
    var soundURI by PreferenceDelegate<String>(SOUND_URI)
    lateinit var sound: Uri
        private set
    var login by PreferenceDelegate<String>(LOGIN)
    var password by PreferenceDelegate<String>(PASSWORD)

    var visibleDistance by PreferenceDelegate<Int>(VISIBLE_DISTANCE)
    var alarmDistance by PreferenceDelegate<Int>(ALARM_DISTANCE)

    var showAccidents by PreferenceDelegate<Boolean>(IS_SHOW_ACCIDENT)
    var showBreaks by PreferenceDelegate<Boolean>(IS_SHOW_BREAK)
    var showSteal by PreferenceDelegate<Boolean>(IS_SHOW_STEAL)
    var showOther by PreferenceDelegate<Boolean>(IS_SHOW_OTHER)

    var vibration by PreferenceDelegate<Boolean>(VIBRATION)

    var hoursAgo by PreferenceDelegate<Int>(HOURS_AGO)

    var maxNotifications by PreferenceDelegate<Int>(MAX_NOTIFICATIONS)

    var isTester by PreferenceDelegate<Boolean>(IS_TESTER)

    private var latitude by PreferenceDelegate<Float>(LATITUDE)
    private var longitude by PreferenceDelegate<Float>(LONGITUDE)

    var savedLatLng
        get() = LatLng(latitude.toDouble(), longitude.toDouble())
        set(latLng) {
            latitude = latLng.latitude.toFloat()
            longitude = latLng.longitude.toFloat()
        }

    var authType by PreferenceDelegate<String>(AUTH_TYPE)

    var vkToken by PreferenceDelegate<String>(VK_TOKEN)

    var googleAccount by PreferenceDelegate<String>(GOOGLE_ACCOUNT)

    var googleName by PreferenceDelegate<String>(GOOGLE_NAME)

    init {
        initSound(MyApp.context)
        val oldVersion = appVersion
        appVersion = MyApp.context.packageManager.getPackageInfo(MyApp.context.packageName, 0).versionCode
        if (oldVersion != appVersion) MyApp.firstStart = true
    }

    private const val DEFAULT_LATITUDE = 55.752295f
    private const val DEFAULT_LONGITUDE = 37.622735f
    private const val DEFAULT_SHOW_DISTANCE = 200
    private const val DEFAULT_ALARM_DISTANCE = 20
    private const val DEFAULT_MAX_NOTIFICATIONS = 3
    private const val DEFAULT_MAX_AGE = 24
    private const val DEFAULT_VIBRATION = true
    private const val DEFAULT_DO_NOT_DISTURB = false
    private const val DEFAULT_SHOW_TYPE = true
    private const val DEFAULT_TESTER = false
    private const val DEFAULT_AUTH_TYPE = "none"

    private enum class Stored(val key: String, val default: Any) {
        IS_SHOW_ACCIDENT("mc.show.acc", DEFAULT_SHOW_TYPE),
        IS_SHOW_BREAK("mc.show.break", DEFAULT_SHOW_TYPE),
        IS_SHOW_STEAL("mc.show.steal", DEFAULT_SHOW_TYPE),
        IS_SHOW_OTHER("mc.show.other", DEFAULT_SHOW_TYPE),
        VISIBLE_DISTANCE("mc.distance.show", DEFAULT_SHOW_DISTANCE),
        ALARM_DISTANCE("mc.distance.alarm", DEFAULT_ALARM_DISTANCE),
        DO_NOT_DISTURB("do.not.disturb", DEFAULT_DO_NOT_DISTURB),
        HOURS_AGO("hours.ago", DEFAULT_MAX_AGE),
        MAX_NOTIFICATIONS("notifications.max", DEFAULT_MAX_NOTIFICATIONS),
        VIBRATION("use.vibration", DEFAULT_VIBRATION),
        LATITUDE("savedlat", DEFAULT_LATITUDE),
        LONGITUDE("savedlng", DEFAULT_LONGITUDE),
        ON_WAY("mc.onway", 0),
        APP_VERSION("mc.app.version", 0),
        SOUND_TITLE("mc.notification.sound.title", "default system"),
        SOUND_URI("mc.notification.sound", ""),
        LOGIN("mc.login", ""),
        PASSWORD("mc.password", ""),
        IS_TESTER("tester", DEFAULT_TESTER),
        AUTH_TYPE("authType", DEFAULT_AUTH_TYPE),
        VK_TOKEN("vkToken", ""),
        GOOGLE_ACCOUNT("googleAccount", ""),
        GOOGLE_NAME("googleName", "");
    }

    init {
        values()
                .filterNot { preferences.contains(it.key) }
                .forEach { setValue(it, it.default) }
    }

    //todo refactor sound settings
    fun setSound(title: String, uri: Uri) {
        soundTitle = title
        soundURI = uri.toString()
    }

    fun initSound(context: Context) {
        sound = if (soundURI == "default") {
            RingtoneManager.getActualDefaultRingtoneUri(context, RingtoneManager.TYPE_NOTIFICATION)
        } else
            Uri.parse(soundURI)
    }

    fun setDefaultSoundAlarm() {
        soundTitle = "default system"
        soundURI = "default"
    }

    fun isEnabled(type: Type): Boolean = when (type) {
        Type.BREAK                                               -> showBreaks
        Type.MOTO_AUTO, Type.MOTO_MOTO, Type.MOTO_MAN, Type.SOLO -> showAccidents
        Type.STEAL                                               -> showSteal
        Type.OTHER                                               -> showOther
        Type.USER                                                -> true
    }

    //todo wtf!?
    fun getPreferenceName(preference: String): String = when (preference) {
        "hoursAgo"         -> "hours.ago"
        "showAcc"          -> "mc.show.acc"
        "showBreak"        -> "mc.show.break"
        "showSteal"        -> "mc.show.steal"
        "showOther"        -> "mc.show.other"
        "distanceShow"     -> "mc.distance.show"
        "distanceAlarm"    -> "mc.distance.alarm"
        "maxNotifications" -> "notifications.max"
        "useVibration"     -> "use.vibration"
        "isTester"         -> "tester"
        else               -> "unknown"
    }

    private fun setValue(stored: Stored, value: Any) {
        preferences.edit().apply {
            when (stored.default) {
                is Int     -> putInt(stored.key, value as Int)
                is Boolean -> putBoolean(stored.key, value as Boolean)
                is String  -> putString(stored.key, value as String)
                is Float   -> putFloat(stored.key, value as Float)
            }
        }.apply()
    }

    private class PreferenceDelegate<T>(private val stored: Stored) : ReadWriteProperty<Preferences, T> {

        @Suppress("UNCHECKED_CAST")
        override fun getValue(thisRef: Preferences, property: KProperty<*>): T =
                when (stored.default) {
                    is String  -> preferences.getString(stored.key, stored.default) as T
                    is Boolean -> preferences.getBoolean(stored.key, stored.default) as T
                    is Int     -> preferences.getInt(stored.key, stored.default) as T
                    is Float   -> preferences.getFloat(stored.key, stored.default) as T
                    else       -> throw TypeCastException("Wrong property type")
                }

        override fun setValue(thisRef: Preferences, property: KProperty<*>, value: T) {
            setValue(stored, value as Any)
        }
    }
}